graveler/committed/sstable/batch.go

```diff
@@ -0,0 +1,68 @@
+package sstable
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/treeverse/lakefs/graveler/committed"
+
+	"go.uber.org/atomic"
+)
+
+type BatchCloser struct {
+	results []committed.WriteResult
+	err     atomic.Error
+
+	wg sync.WaitGroup
+
+	// resultsLock locks any access to the accumulated results
+	resultsLock sync.Mutex
+}
+
+var (
+	errMultipleWaitCalls = errors.New("wait has already been called")
+)
+
+// CloseWriterAsync adds Writer instance for the BatchWriterCloser to handle.
+// Any writes executed to the writer after this call are not guaranteed to succeed.
+// If Wait() has already been called, returns an error.
+func (bc *BatchCloser) CloseWriterAsync(w committed.Writer) error {
+	err := bc.err.Load()
+	if err != nil {
+		// don't accept new writers if previous error occurred
+		return err
+	}
+
+	bc.wg.Add(1)
+	go bc.closeWriter(w)
+
+	return nil
+}
+
+func (bc *BatchCloser) closeWriter(w committed.Writer) {
+	defer bc.wg.Done()
+	res, err := w.Close()
+	if err != nil {
+		if bc.err.Load() == nil {
+			// keeping first error is enough
+			bc.err.Store(err)
+		}
```

Comment:
This is still a race, and does not keep the first error (but it does keep _an_ error).  You would need a `Swap` operation like [`Uint64.Swap](https://godoc.org/go.uber.org/atomic#Uint64.Swap), probably using `unsafe.Pointer`s and [`atomic.SwapPointer`](https://golang.org/pkg/sync/atomic/#SwapPointer).
But again, I doubt that it matters.  You could just change the semantics and
```suggestion
	if err != nil {
		// keeping last error is enough
		bc.err.Store(err)
	}
```

Reply:
Or _don't use atomics!  Why bother to speed up the error-handling case when the good path (below) has a lock?